lib/stoplight/data_store/redis.rb

```diff
@@ -17,10 +17,72 @@ module DataStore
     class Redis < Base
       KEY_SEPARATOR = ":"
       KEY_PREFIX = %w[stoplight v5].join(KEY_SEPARATOR)
+      # This Lua script implements a notification deduplication mechanism:
+      # 1. It checks if the current color transition (from_color → to_color) matches the previously recorded one
+      # 2. If it's the same transition, returns 0 (don't notify) to prevent duplicate notifications
+      # 3. If it's a different transition (or first time), it:
+      #    - Records the new transition with a timestamp
+      #    - Sets an expiration time on the record (allowing notifications to repeat after TTL as a fail-safe mechanism)
+      #    - Returns 1 (proceed with notification)
+      #
+      # This ensures that when multiple servers detect the same transition simultaneously,
+      # only one notification is sent, while still allowing future transitions to trigger notifications.
+      #
+      # This script is executed on redis and it's guaranteed that the operations are atomic.
+      NOTIFICATION_DEDUPLICATION_SCRIPT = <<~LUA
+        local last_notification_key = KEYS[1]
+
+        local light_name = ARGV[1]
+        local from_color = ARGV[2]
+        local to_color = ARGV[3]
+        local ttl = tonumber(ARGV[4])
+
+        local prev_transition = redis.call('HMGET', last_notification_key, 'from_color', 'to_color')
+        local prev_from_color, prev_to_color = unpack(prev_transition)
+
+        if prev_from_color == from_color and prev_to_color == to_color then
+          return 0
+        else
+          redis.call('HSET', last_notification_key, 'from_color', from_color, 'to_color', to_color)
+          redis.call('EXPIRE', last_notification_key, ttl)
+          return 1
+        end
+      LUA
+      private_constant :NOTIFICATION_DEDUPLICATION_SCRIPT
+
+      RECORD_FAILURE_SCRIPT = <<~LUA
+        local failures_key = KEYS[1]
+        local current_time = tonumber(ARGV[1])
+        local window_size = tonumber(ARGV[2])
+        local threshold = tonumber(ARGV[3])
+        local failure_json = ARGV[4]
+        
+        -- Add new failure
+        redis.call('ZADD', failures_key, current_time, failure_json)
+        
+        -- Calculate window boundaries
+        local window_start = current_time - window_size
+        
+        -- Remove failures outside time window
+        redis.call('ZREMRANGEBYSCORE', failures_key, 0, window_start)
+          
+        -- Keep at most threshold failures (remove oldest)
+        redis.call('ZREMRANGEBYRANK', failures_key, 0, -threshold - 1)
+        
+        -- Count only failures within current window
+        return redis.call('ZCOUNT', failures_key, window_start, '+inf')
+      LUA
+      private_constant :RECORD_FAILURE_SCRIPT
 
       # @param redis [::Redis, ConnectionPool<::Redis>]
       def initialize(redis)
         @redis = redis
+        @notification_deduplication_script_sha = @redis.then do |client|
```

Comment:
Upload scripts during initialization, allowing for later referencing by SHA. This significantly reduces the command payload when executing such scripts.